Changes for more model flexibility

ccabello99 opened the pull request:
Included the concept of adding JSON files for flexibility in parameter changes irrespective of specific model. This is a rough draft and refinements are needed

ccabello99 committed:
Changes for more model flexibility

BenediktBurger commented:
It seems, that you based your work on an old branch. Would you mind rebasing your branch on the most current main branch?

ccabello99 committed:
Updating to new framework

ccabello99 committed:
Used similar approach as for IS/Pixelink plugins

ccabello99 committed:
Fixed trigger save setting & added h5 save support

BenediktBurger commented:
Please merge the current master branch into your branch (or rebase onto master) as you revert some changes, which we made since you started to work on your branch.

If you have questions how to do it, don't hesitate to ask.

ccabello99 committed:
Typo in plugin_info.toml

ccabello99 commented:
Hello Benedikt,

I did: git merge upstream/main, and it outputs: "Already up to date."

Am I missing something ?

BenediktBurger commented on src/pymodaq_plugins_basler/hardware/basler.py:
```diff
@@ -34,14 +27,23 @@ class DartCamera:
+        self.device_info = info
+
+        # Default place to look for saved device configuration
+        if platform.system() == 'Windows':
+            self.default_device_state_path = os.path.join(
+                os.environ.get('PROGRAMDATA'), '.pymodaq', f'{self.model_name}_config.pfs'
+            )
+        else:
+            self.default_device_state_path = os.path.join(
+                '/etc', '.pymodaq', f'{self.model_name}_config.pfs'
+            )
```
Do you know, that you can store/retrieve things with QSettings from the Qt package?

BenediktBurger replied:
If you want to use pymodaqs config system, you should use that instead of using your own (or is this how PyMoDAQ does it? I don't know it too well).

You're right, that it might make sense to have it stored together with other PyMoDAQ data.

In general, QSettings makes it easier to save settings, as you do not have to worry about the location (it is in the registry in Windows and in /etc in Linux/Mac) or file path. You can always retrieve the documentation, wherever your program is.

ccabello99 replied:
As far as I was aware, PyMoDAQ is using the /.pymodaq directory to store config files in general, which is in /etc in Linux and /ProgramData in Windows. 

For this case, where someone saves a device configuration state from the plugin, it did make sense to me to rather default to this same directory for saving and loading. If someone happens to save the nodemap .pfs file from pylon Viewer, for example, then we are simply asking that they move this file to the .pymodaq directory if they want to use it for the plugin.

BenediktBurger commented on src/pymodaq_plugins_basler/resources/VERSION:
that file is not according to the new plugin format anymore and should not be here.

ccabello99 replied:
Done !

BenediktBurger commented on hatch_build.py:
why do you delete that file and add plugin_info.toml etc?

If you need them for your local setup (as the comment in plugin_info suggests) you can keep them locally, but do not check them in into git.

BenediktBurger reviewed (commented):
A few generic comments.

@rgeneaux has to take a look at the content itself.

ccabello99 committed:
Updated build config, added LECO logging concept

ccabello99 committed:
Updated README, and small refactor/syntax changes

ccabello99 commented:
Hi Romain,

Thank you for the comment. This idea came about from discussions with @Semptum and @smartslava by trying to find a device agnostic approach to plugin operation. I have updated the README to detail the idea and procedure behind the configuration files.

It should suffice for most camera models to simply change config_<model_name>.json with the correct model name of the device. However, this approach allows some flexibility to add/remove parameters as wanted, and removes the trouble of saving specific device class attributes depending on whether its gain value is accessed via GainRaw or GainAbs, for example. This need only be changed in the json file, and the rest will be handled automatically. 

If a user would like to keep only very basic functionalities, for instance, they only care to be able to control exposure time and gain. Then, they can simply delete all the other parameters in the json file, and it should work just fine. Adding parameters is a bit more complicated right now since other than basic numeric parameters, there will likely need to be some additional handling in the update_params_ui method, which I added, and the normal commit_settings method.

Please let me know if you have any other concerns or questions.

Christian

ccabello99 committed:
Small typo fix and extra comment

ccabello99 committed:
Fixed small device state save/load bug

ccabello99 committed:
Use uuid7 instead of ulid in LECO metadata

ccabello99 committed:
Small typo fixes

BenediktBurger commented on pyproject.toml:
```diff
@@ -14,7 +14,9 @@ description = 'PyMoDAQ plugins for cameras of Basler'
     'pymodaq>=4.4.7',
     'numpy',  # for Basler camera
-    'pypylon',
+    'pypylon'
+    'imageio',
+    'uuid7'
```
two notes:
1. you did not add a comma after pypylon
2. https://pypi.org/project/uuid7/ implemented an earlier version of UUIDv7, which generates dates after the year 2160 or so, so you should consider using another library, see https://github.com/stevesimmons/uuid7/issues/1. I used another package, see https://github.com/pymeasure/pyleco/issues/74

ccabello99 replied:
1. Thanks for catching the typo.
2. Thanks again for the tip. I've made the change to use the uuid6 package instead, but still send str(uuid6.uuid7()) in the metadata.

ccabello99 committed:
Typo and metadata fixes and uuid7 -> uuid6 package

ccabello99 committed:
Fix typos

ccabello99 committed:
Include h5py in dependencies

ccabello99 committed:
Fuzziness metadata in ms and fix temp. monitor bug

ccabello99 committed:
Only emit metadata when in trigger saving mode

ccabello99 committed:
Fix typo

ccabello99 committed:
Fixed small metadata setting and trigger saving bugs

ccabello99 committed:
Get leco basepath with qsettings

ccabello99 committed:
Fix metadata and leco basepath and fix saving in emit_data_callback

ccabello99 committed:
Fix some saving h5 issues

ccabello99 committed:
Turn off saving on detector close, remove setup.py

ccabello99 committed:
Typo

ccabello99 committed:
Extra metadata for version control of h5 formatting

ccabello99 committed:
Moved LECO operations to new 2D viewer class to keep original generic and always include device info in parameter tree

ccabello99 committed:
Change save location parameter type & wrap get attr in try/except

ccabello99 committed:
Add device info parameter group to param tree by default

ccabello99 committed:
set self.save_frame to false before we stop trigger

ccabello99 committed:
Only handle metadata if self.save_frame

ccabello99 committed:
self.save_frame false before turning off trigger

ccabello99 committed:
Safety measures to avoid saving while not triggered

ccabello99 committed:
Fixes for compatability with other camera models

ccabello99 committed:
Fix filename when saving is not h5

ccabello99 committed:
Updated pyproject

ccabello99 committed:
Default config file creation if not found & cleanup

ccabello99 commented:
Hi @rgeneaux,

Thanks for the comments and excuse the delay. I found option two to also be a nice compromise and maybe a bit more practical for most users, so I finished a working implementation of that. 

It does as you suggest. If the config file is not found in the expected directory, "C:\ProgramData\.pymodaq" for Windows and "/etc/.pymodaq/" for Linux, then a popup window will be raised asking if the user would like to create one. 

If yes, a basic config file for the camera model will be created for access to exposure and gain settings. It will then let the user know its filepath and that it can be modified to add/remove parameters. 

If no, another popup will be raised letting them know that they cannot access the camera parameters until a valid config is available, but then the camera will continue initialization anyways and will be usable for grabbing images.

There can still be refinements as I currently get some warnings from the main thread when the popup is being executed as :
QBasicTimer::stop: Failed. Possibly trying to stop from a different thread
QBackingStore::endPaint() called with active painter; did you forget to destroy it or call QPainter::end() on it?

However, it does not seem to lead to any segfaults or other errors for me. I could not identify a solution to this as of now, but just so that you are aware.

ccabello99 committed:
Safer message box popup for default config creation in Windows

ccabello99 committed:
Update README

ccabello99 committed:
Update dependencies

ccabello99 committed:
Update exposure and gain value in GUI if auto; also, only use auto mode as continuous

ccabello99 committed:
Remove BaslerWithLECO class because it is a bit specific to our needs and not generalized

rgeneaux commented:
Thanks @ccabello99 for the nice PR!

BenediktBurger commented:
Nice that this long story reached its goal.
